src/Workspaces/Core/Portable/Utilities/AsyncLazy`1.cs

```diff
@@ -474,7 +474,8 @@ private Task<T> GetCachedValueAndCacheThisValueIfNoneCached_NoLock(Task<T> task)
                 if (_cacheResult && task.Status == TaskStatus.RanToCompletion)
                 {
                     // Hold onto the completed task. We can get rid of the computation functions for good
-                    _cachedResult = task;
+                    Volatile.Write(ref _cachedResult, task);
```

Comment:
Although you do a raise a good question about whether we have a write reordering bug there. I'd appreciate it in a separate PR either way, but I can't say at first glance this is wrong...